src/if_statements.py

```diff
@@ -355,228 +343,240 @@ def emit_goto_or_early_return(context: Context, target: Node, body: Body) -> Non
+    assert else_node is not None
+
+    # Check that all edges point "forward" to other nodes in the if statement
+    # and translate this DAG of nodes into a dict we can easily modify
+    allowed_nodes = set(chained_cond_nodes) | {if_node, else_node}
+    node_cond_edges: Dict[ConditionalNode, Tuple[Condition, Node, Node]] = {}
+    for node in chained_cond_nodes:
+        if (
+            node.conditional_edge not in allowed_nodes
+            or node.fallthrough_edge not in allowed_nodes
+        ):
+            # Not a valid set of chained_cond_nodes
+            return None
+        allowed_nodes.remove(node)
+
+        block_info = node.block.block_info
+        assert block_info
+        if node is start_node:
             # The first condition in an if-statement will have unrelated
             # statements in its to_write list, which our caller will already
             # have emitted. Avoid emitting them twice.
-            conditions.append(branch_condition)
+            cond = block_info.branch_condition
         else:
-            # Include the statements in the condition by using a comma expression.
-            conditions.append(gather_any_comma_conditions(block_info))
-        condition_nodes.append(curr_node)
+            # Otherwise, these statements will be added to the condition
+            cond = gather_any_comma_conditions(block_info)
+
+        node_cond_edges[node] = (cond, node.conditional_edge, node.fallthrough_edge)
+
+    # Iteratively (try to) reduce the nodes into a single condition
+    #
+    # This is done through a process similar to "Rule T2" used in interval analysis
+    # of control flow graphs, see ref. slides 17-21 of:
+    # http://misailo.web.engr.illinois.edu/courses/526-sp17/lec1.pdf
+    #
+    # We have already ensured that all edges point forward (no loops), and there
+    # are no incoming edges to internal nodes from outside the chain.
+    #
+    # Pick the first pair of nodes which form one of the 4 possible reducible
+    # subgraphs, and then "collapse" them together by combining their conditions
+    # and adjusting their edges. This process is repeated until no more changes
+    # are possible, and is a success if there is exactly 1 condition left.
+    while True:
+        # Calculate the parents for each node in our subgraph
+        node_parents: Dict[ConditionalNode, List[ConditionalNode]] = {
+            node: [] for node in node_cond_edges
+        }
+        for node in node_cond_edges:
+            for child in node_cond_edges[node][1:]:
+                if child not in (if_node, else_node):
+                    assert isinstance(child, ConditionalNode)
+                    node_parents[child].append(node)
+
+        # Find the first pair of nodes which form a reducible pair: one will always
+        # be the *only* parent of the other.
+        # Note: we do not include `if_node` or `else_node` in this search
+        for child, parents in node_parents.items():
```

Comment:
I originally thought that I could make this a bit more efficient by iterating through the nodes _backwards_, so that I wouldn't have to recalculate `node_parents` after every change. But, this produces "right heavy" or irregular `BinaryOp` trees.

There are a few places that assume "left heavy" `BinaryOp` trees, so I didn't think the perf was worth the code complexity.

Reply:
The algorithm is O(n^4) in the worst-case (lots of consecutive ConditionalNode's that don't fold into a joined conditions, but with dominators that still join them together into a single group), but it's probably not that bad in practice. Calculating `node_parents` incrementally and fixing up irregular BinaryOp's once we've found a solution (not too hard recursively) would make it O(n^3), but I think it's fair to put that off until it becomes a problem.

Changing the &&-with-comma-conditions logic to construct nested if's instead of just dropping the last node and trying again would also help perf in practice, but not asymptotically.